servlet/src/main/java/com/linecorp/armeria/server/servlet/ServletBuilder.java

```diff
@@ -0,0 +1,180 @@
+ * under the License.
+ */
+package com.linecorp.armeria.server.servlet;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+import java.util.Objects;
+
+import javax.annotation.Nullable;
+import javax.servlet.Filter;
+import javax.servlet.ServletException;
+import javax.servlet.ServletRegistration;
+import javax.servlet.http.HttpServlet;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.linecorp.armeria.server.ServerBuilder;
+import com.linecorp.armeria.server.servlet.util.MimeMappings;
+
+/**
+ * A builder class which creates a new {@link DefaultServletContext} instance.
+ */
+public class ServletBuilder {
+    private static final Logger logger = LoggerFactory.getLogger(ServletBuilder.class);
+    private final DefaultServletContext servletContext;
+    private final ServerBuilder serverBuilder;
+    private final String contextPath;
+
+    /**
+     * Constructor.
+     */
+    public ServletBuilder(ServerBuilder serverBuilder, @Nullable String contextPath) {
+        requireNonNull(serverBuilder, "serverBuilder");
+        if (Objects.isNull(contextPath)) {
+            contextPath = "";
+        }
+        if (!contextPath.isEmpty()) {
+            checkArgument(contextPath.charAt(0) == '/' && !contextPath.substring(1).contains("/"),
+                          "Context path have to start with / and contains only one /", contextPath);
+        }
+        servletContext = new DefaultServletContext(contextPath);
+        this.contextPath = contextPath;
+        this.serverBuilder = serverBuilder;
+    }
+
+    /**
+     * Creates a new {@link ServerBuilder}.
+     */
+    public ServerBuilder build() {
+        return serverBuilder;
+    }
+
+    /**
+     * Add a servlet.
+     */
+    public ServletBuilder servlet(String path, HttpServlet httpServlet) {
+        requireNonNull(path, "path");
+        requireNonNull(httpServlet, "httpServlet");
+        return addServlet(path, httpServlet);
+    }
+
+    /**
+     * Add a servlet.
+     */
+    public ServletBuilder servlet(String path, String servletClass) {
+        requireNonNull(path, "path");
+        requireNonNull(servletClass, "servletClass");
+        return addServlet(path, servletClass);
+    }
+
+    private ServletBuilder addServlet(String path, Object servlet) {
+        requireNonNull(path, "path");
+        requireNonNull(servlet, "servlet");
+        if (!"".equals(contextPath)) {
+            path = "/".equals(path) ? contextPath : contextPath + path;
+        }
+        ServletRegistration sd = null;
+        if (servlet instanceof String) {
+            sd = servletContext.addServlet(path, (String) servlet);
+        } else {
+            sd = servletContext.addServlet(path, (HttpServlet) servlet);
+        }
+        sd.addMapping(path);
+        serverBuilder.service(path, new DefaultServletService(servletContext));
```

Comment:
I think we can just add one `DefaultServletService` when `ServletBuilder.build()` is called.
```
serverBuilder.serviceUnder(contextPath, new DefaultServletService(servletContext));
```
Because we use `context.getRequestDispacher(path)` in `DefaultServletService` to find the corresponding `Servlet`.

Reply:
Probably not, because user can add servlet root path "/" or "" (It is also context path) : `sb.servlet("/", new HomeServlet())` or `sb.servlet("", new HomeServlet())` . I think user should decide whether to implement servlet root path or not.

Reply:
> sb.servlet("/", new HomeServlet())
or
sb.servlet("", new HomeServlet())

What's the URL that a user connects to the servlet in those cases?

Reply:
URL is `http://host:port/contextPath` that is initialized in DefaultServletContext. 

	`new DefaultServletContext(contextPath)`

ServletBuilder.java

```
    public ServletBuilder(ServerBuilder serverBuilder, String contextPath) {
        requireNonNull(serverBuilder, "serverBuilder");
        requireNonNull(contextPath, "contextPath");
        if (!contextPath.isEmpty()) {
            checkArgument(contextPath.charAt(0) == '/' &&
                           contextPath.charAt(contextPath.length() - 1) != '/',
                          "Context path format is /x/y/z", contextPath);
        }
        servletContext = new DefaultServletContext(contextPath);
        this.contextPath = contextPath;
        this.serverBuilder = serverBuilder;
    }
```

Reply:
What I meant was what's the difference between URLs when `contextPath` is `"/"` and `""`?
Isn't it `http://host:port/` in both cases?

Reply:
Armeria server not support empty service path (""). But servlet support empty path. If we access `http://host:port/`, server return `http://host:port` => Http response status is not 200(ok), it is 307 (redirect) because URL was changed. If we should support both cases "/" and "", Http response status alway be 200 (ok).

Reply:
> Armeria server not support empty service path ("")

Right, Armeria does not support the empty path so a servlet on Armeria couldn't have empty path mapping. Isn't that right?

I was thinking just adding `/` after `contextPath` when we call
`serverBuilder.serviceUnder(...);` because the `contextPath` does not end with `/` 
so that we can just handle the servlets by just adding one `DefaultServletService`.

When `contextPath` is `/foo`:
- if a user adds a servlet too `/bar`, then the url is `http://host:port/foo/bar`
When `contextPath` is `""`:
- if a user adds a servlet too `/baz`, then the url is `http://host:port/baz`
When `contextPath` is `/`:
- This is an exception.

What do you think?

Reply:
Yes, it's like you explained. I will fix this exception. Do you suggest any solution for this problem

Reply:
I think add DefaultServletService is not perfect solution. It is very nice if it is always default return Not found with path `/` , and users add a servlet to `/` if they need.